rust/otap-dataflow/crates/state/src/pipeline_status.rs

```diff
@@ -0,0 +1,365 @@
+    /// Coarse phase synthesized from all per-core phases.
+    phase: PipelineAggPhase,
+
+    /// Per-core details to aid debugging and aggregation.
+    pub(crate) cores: HashMap<CoreId, PipelineRuntimeStatus>,
+
+    #[serde(skip)]
+    health_policy: HealthPolicy,
+}
+
+impl PipelineStatus {
+    pub(crate) fn new(health_policy: HealthPolicy) -> Self {
+        Self {
+            phase: PipelineAggPhase::Unknown,
+            cores: HashMap::new(),
+            health_policy,
+        }
+    }
+
+    /// Returns the current aggregated phase of the pipeline.
+    #[must_use]
+    pub fn phase(&self) -> &PipelineAggPhase {
+        &self.phase
+    }
+
+    /// Returns the current per-core status map.
+    #[must_use]
+    pub fn per_core(&self) -> &HashMap<CoreId, PipelineRuntimeStatus> {
+        &self.cores
+    }
+
+    fn counts(&self) -> AggregateCounts {
+        let mut agg = AggregateCounts::default();
+        for c in self.cores.values() {
+            match &c.phase {
+                PipelinePhase::Pending => agg.pending += 1,
+                PipelinePhase::Starting => agg.starting += 1,
+                PipelinePhase::Running => agg.running += 1,
+                PipelinePhase::Updating => agg.updating += 1,
+                PipelinePhase::RollingBack => agg.rolling_back += 1,
+                PipelinePhase::Draining => agg.draining += 1,
+                PipelinePhase::Stopped => agg.stopped += 1,
+                PipelinePhase::Failed(r) => {
+                    agg.failed += 1;
+                    *agg.failed_reasons.entry(*r).or_insert(0) += 1;
+                }
+                PipelinePhase::Deleting(m) => {
+                    agg.deleting += 1;
+                    if matches!(m, DeletionMode::Forced) {
+                        agg.forced_deletes += 1;
+                    }
+                }
+                PipelinePhase::Deleted => agg.deleted += 1,
+                PipelinePhase::Rejected(r) => {
+                    agg.rejected += 1;
+                    *agg.rejected_reasons.entry(*r).or_insert(0) += 1;
+                } // PipelinePhase::Unknown => agg.unknown += 1,
+            }
+        }
+        agg
+    }
+
+    /// Infer the aggregated phase.
+    ///
+    /// Choose a single, meaningful headline phase from the per-core counts.
+    /// Precedence (highest first): Deleted, Deleting, Failed, Rejected, RollingBack, Updating,
+    /// Draining, RunningAll/RunningDegraded, Starting, StoppedAll/StoppedPartial.
+    pub fn infer_agg_phase(&mut self) {
+        let count = self.counts();
+        let total = count.total();
+        let active = count.active();
+        let forced = count.forced_deletes > 0;
+        let top_fail = count
+            .failed_reasons
+            .iter()
+            .max_by_key(|(_, n)| *n)
+            .map(|(r, _)| *r);
+        let top_reject = count
+            .rejected_reasons
+            .iter()
+            .max_by_key(|(_, n)| *n)
+            .map(|(r, _)| *r);
```

Comment:
nit: it might be a slightly better performance to compute these where we need them? e.g. we only need `top_fail` in the block below near 112 where we're setting `self.phase = PipelineAggPhase::Failed`